File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/SpecialPower/CleanupAreaPower.cpp
// FILE: CleanupAreaPower.cpp /////////////////////////////////////////////////
//-----------------------------------------------------------------------------
//                                                                          
//                       Electronic Arts Pacific.                          
//                                                                          
//                       Confidential Information                           
//                Copyright (C) 2002 - All Rights Reserved                  
//                                                                          
//-----------------------------------------------------------------------------
//
//	Created:	September 2002
//
//	Author:		Kris Morness
//	
//  Makes use of the cleanup hazard update by augmenting the cleanup range 
//  until there is nothing left to cleanup at which time it goes idle.
//-----------------------------------------------------------------------------
///////////////////////////////////////////////////////////////////////////////

//-----------------------------------------------------------------------------
#include "PreRTS.h"	// This must go first in EVERY cpp file int the GameEngine

#include "Common/Player.h"
#include "Common/ThingTemplate.h"
#include "Common/Xfer.h"
#include "GameLogic/Module/CleanupAreaPower.h"
#include "GameLogic/Module/CleanupHazardUpdate.h"
#include "GameLogic/Object.h"

//-------------------------------------------------------------------------------------------------
CleanupAreaPowerModuleData::CleanupAreaPowerModuleData()
{
	m_cleanupMoveRange = 0.0;
} 

//-------------------------------------------------------------------------------------------------
void CleanupAreaPowerModuleData::buildFieldParse(MultiIniFieldParse& p)
{
	SpecialPowerModuleData::buildFieldParse( p );

	static const FieldParse dataFieldParse[] = 
	{
		{ "MaxMoveDistanceFromLocation",			INI::parseReal, NULL, offsetof( CleanupAreaPowerModuleData, m_cleanupMoveRange ) },
		{ 0, 0, 0, 0 } 
	};
	p.add(dataFieldParse);

}  // end buildFieldParse

//-------------------------------------------------------------------------------------------------
CleanupAreaPower::CleanupAreaPower( Thing *thing, const ModuleData* moduleData ) : SpecialPowerModule( thing, moduleData )
{

}

//-------------------------------------------------------------------------------------------------
CleanupAreaPower::~CleanupAreaPower()
{

}

//-------------------------------------------------------------------------------------------------
void CleanupAreaPower::doSpecialPowerAtLocation( const Coord3D *loc, Real angle, UnsignedInt commandOptions )
{
	if (getObject()->isDisabled())
		return;

	Object *obj = getObject();
	const CleanupAreaPowerModuleData *data = getCleanupAreaPowerModuleData();

	static NameKeyType key_CleanupHazardUpdate = NAMEKEY( "CleanupHazardUpdate" );
	CleanupHazardUpdate *update = (CleanupHazardUpdate*)obj->findUpdateModule( key_CleanupHazardUpdate );
	if( update )
	{
		update->setCleanupAreaParameters( loc, data->m_cleanupMoveRange );
	}
	else
	{
		//This case will only happen should the ambulance not have a cleanuphazard update module.
		DEBUG_CRASH( ("%s is attempting to use CleanupAreaPower, but requires a CleanupHazardUpdate module to work!", obj->getTemplate()->getName().str() ) );
	}
}


//-------------------------------------------------------------------------------------------------
void CleanupAreaPower::crc( Xfer *xfer )
{

	// extend base class
	SpecialPowerModule::crc( xfer );

}  // end crc

//-------------------------------------------------------------------------------------------------
// Xfer method
// Version Info:
// 1: Initial version 
//-------------------------------------------------------------------------------------------------
void CleanupAreaPower::xfer( Xfer *xfer )
{

	// version
	XferVersion currentVersion = 1;
	XferVersion version = currentVersion;
	xfer->xferVersion( &version, currentVersion );

	// extend base class
	SpecialPowerModule::xfer( xfer );

}  // end xfer

//-------------------------------------------------------------------------------------------------
void CleanupAreaPower::loadPostProcess( void )
{

	// extend base class
	SpecialPowerModule::loadPostProcess();

}  // end loadPostProcess

File: GeneralsRust/Code/GameEngine/GameLogic/src/object/special_powers/cleanup_area_power.rs
//! CleanupAreaPower
//!
//! Port of CleanupAreaPower.h and CleanupAreaPower.cpp
//! Author: Kris Morness, September 2002 (C++), Rust Port
//!
//! Makes use of the CleanupHazardUpdate module by augmenting the cleanup range
//! until there is nothing left to clean up, at which time it goes idle.
//! Used by the Ambulance to clean mines and other hazards in an area.
//!
//! Key behavior:
//! - doSpecialPowerAtLocation: delegates to CleanupHazardUpdate::setCleanupAreaParameters()
//! - The CleanupHazardUpdate module handles the actual movement and cleanup logic

use std::sync::Arc;

use game_engine::common::ini::{FieldParse, INIError, INI};
use game_engine::common::name_key_generator::NameKeyGenerator;
use game_engine::common::system::Snapshotable;
use game_engine::common::thing::module::{Module, ModuleData, NameKeyType};

use crate::common::{Coord3D, ObjectID, Real};
use crate::helpers::TheGameLogic;
use crate::modules::{BehaviorModuleInterface, CleanupHazardUpdateInterface};
use crate::object::special_power_module::SpecialPowerModuleData;

/// Module data for CleanupAreaPower.
/// Matches C++ CleanupAreaPowerModuleData.
#[derive(Debug, Clone)]
pub struct CleanupAreaPowerModuleData {
    pub base: SpecialPowerModuleData,
    /// Maximum move distance from the cleanup location.
    /// Matches C++ m_cleanupMoveRange.
    pub cleanup_move_range: Real,
}

impl Default for CleanupAreaPowerModuleData {
    fn default() -> Self {
        Self {
            base: SpecialPowerModuleData::default(),
            cleanup_move_range: 0.0,
        }
    }
}

impl CleanupAreaPowerModuleData {
    pub fn parse_from_ini(&mut self, ini: &mut INI) -> Result<(), INIError> {
        ini.init_from_ini_with_fields(self, CLEANUP_AREA_POWER_FIELDS)
    }
}

impl ModuleData for CleanupAreaPowerModuleData {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn set_module_tag_name_key(&mut self, key: NameKeyType) {
        self.base.base.set_module_tag_name_key(key);
    }

    fn get_module_tag_name_key(&self) -> NameKeyType {
        self.base.base.get_module_tag_name_key()
    }
}

impl Snapshotable for CleanupAreaPowerModuleData {
    fn crc(&self, _xfer: &mut dyn game_engine::common::system::Xfer) -> Result<(), String> {
        self.base.crc(_xfer)
    }

    fn xfer(&mut self, xfer: &mut dyn game_engine::common::system::Xfer) -> Result<(), String> {
        self.base.xfer(xfer)
    }

    fn load_post_process(&mut self) -> Result<(), String> {
        self.base.load_post_process()
    }
}

/// CleanupAreaPower module.
///
/// Matches C++ CleanupAreaPower which extends SpecialPowerModule.
/// Sets cleanup area parameters on the owning object's CleanupHazardUpdate module.
pub struct CleanupAreaPower {
    module_name_key: NameKeyType,
    data: Arc<CleanupAreaPowerModuleData>,
    owner_object_id: ObjectID,
}

impl CleanupAreaPower {
    pub fn new(
        module_name_key: NameKeyType,
        owner_object_id: ObjectID,
        data: Arc<CleanupAreaPowerModuleData>,
    ) -> Self {
        Self {
            module_name_key,
            data,
            owner_object_id,
        }
    }

    /// Execute cleanup area power at location.
    /// Matches C++ CleanupAreaPower::doSpecialPowerAtLocation().
    pub fn do_special_power_at_location(&self, loc: &Coord3D) -> Result<(), String> {
        // Check disabled
        let Some(owner) = TheGameLogic::find_object_by_id(self.owner_object_id) else {
            return Ok(());
        };
        {
            let Ok(owner_guard) = owner.read() else {
                return Ok(());
            };
            if owner_guard.is_disabled() {
                return Ok(());
            }
        }

        let move_range = self.data.cleanup_move_range;

        // Find the CleanupHazardUpdate module on the owner
        // C++: obj->findUpdateModule("CleanupHazardUpdate")
        let Some(owner) = TheGameLogic::find_object_by_id(self.owner_object_id) else {
            return Ok(());
        };
        let Ok(owner_guard) = owner.read() else {
            return Ok(());
        };

        let Some(module) = owner_guard.find_update_module("CleanupHazardUpdate") else {
            log::debug!("CleanupAreaPower: owner missing CleanupHazardUpdate module");
            return Ok(());
        };

        // Delegate to CleanupHazardUpdate::setCleanupAreaParameters(loc, range)
        // C++: update->setCleanupAreaParameters(loc, data->m_cleanupMoveRange)
        module.with_module_downcast::<crate::object::behavior::cleanup_hazard_update::CleanupHazardUpdate, _, _>(|behavior| {
            CleanupHazardUpdateInterface::set_cleanup_area_parameters(behavior, loc, move_range);
        });

        Ok(())
    }
}

impl Module for CleanupAreaPower {
    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
        self
    }

    fn get_module_name_key(&self) -> NameKeyType {
        self.module_name_key
    }

    fn get_module_tag_name_key(&self) -> NameKeyType {
        self.data.get_module_tag_name_key()
    }

    fn get_module_data(&self) -> &dyn ModuleData {
        self.data.as_ref()
    }
}

impl Snapshotable for CleanupAreaPower {
    fn crc(&self, _xfer: &mut dyn game_engine::common::system::Xfer) -> Result<(), String> {
        // Version 1: Initial version - extends base class only
        Ok(())
    }

    fn xfer(&mut self, xfer: &mut dyn game_engine::common::system::Xfer) -> Result<(), String> {
        // Version 1: Initial version - extends base class only
        let mut version: u8 = 1;
        xfer.xfer_version(&mut version, 1)
            .map_err(|e| format!("CleanupAreaPower xfer version failed: {:?}", e))?;
        Ok(())
    }

    fn load_post_process(&mut self) -> Result<(), String> {
        // Matches C++ CleanupAreaPower::loadPostProcess()
        Ok(())
    }
}

impl BehaviorModuleInterface for CleanupAreaPower {
    fn get_module_name(&self) -> &'static str {
        "CleanupAreaPower"
    }
}

// INI field parsers

fn parse_special_power_template_field(
    _ini: &mut INI,
    data: &mut CleanupAreaPowerModuleData,
    tokens: &[&str],
) -> Result<(), INIError> {
    let token = tokens.first().ok_or(INIError::InvalidData)?;
    let name = crate::common::AsciiString::from(*token);
    data.base.special_power_template =
        Some(crate::object::special_power_template::find_or_create_special_power_template(&name));
    Ok(())
}

fn parse_max_move_distance_from_location(
    _ini: &mut INI,
    data: &mut CleanupAreaPowerModuleData,
    tokens: &[&str],
) -> Result<(), INIError> {
    let token = tokens
        .iter()
        .find(|t| **t != "=")
        .ok_or(INIError::InvalidData)?;
    data.cleanup_move_range = INI::parse_real(token)?;
    Ok(())
}

const CLEANUP_AREA_POWER_FIELDS: &[FieldParse<CleanupAreaPowerModuleData>] = &[
    FieldParse {
        token: "SpecialPowerTemplate",
        parse: parse_special_power_template_field,
    },
    FieldParse {
        token: "MaxMoveDistanceFromLocation",
        parse: parse_max_move_distance_from_location,
    },
];

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cleanup_area_default() {
        let data = CleanupAreaPowerModuleData::default();
        assert_eq!(data.cleanup_move_range, 0.0);
    }

    #[test]
    fn test_cleanup_area_module_name() {
        let data = CleanupAreaPowerModuleData::default();
        let arc_data = Arc::new(data);
        let power = CleanupAreaPower::new(0, 0, arc_data);
        assert_eq!(power.get_module_name(), "CleanupAreaPower");
    }

    #[test]
    fn test_do_special_power_at_location_no_owner() {
        let data = CleanupAreaPowerModuleData::default();
        let arc_data = Arc::new(data);
        let power = CleanupAreaPower::new(0, 0, arc_data);
        // Should return Ok without panicking when owner doesn't exist
        assert!(power
            .do_special_power_at_location(&Coord3D::new(0.0, 0.0, 0.0))
            .is_ok());
    }
}
